### PR TITLE
geyser: add flag `geyser-plugin-snapshot-disabled`

### PR DESCRIPTION
**Pull request by fanatid**
#### Problem

Startup time with an enabled geyser interface is significant and CPU load does not exceed 100%. Disable snapshot on nodes that need geyser but do not need snapshot reduce startup time on 15-20min.

#### Summary of Changes

PR introduces a new flag `--geyser-plugin-snapshot-disabled`, if flag is set then account notifications from snapshot to geyser are ignored and the node starts faster while geyser interface is still enabled.

### CODE REVIEW

File: accounts-db/src/accounts_db/geyser_plugin_utils.rs [hunk 1]
```diff
@@ -39,22 +39,23 @@ impl AccountsDb {
...
     /// multiple times only the last write (with highest write_version) is notified.
     pub fn notify_account_restore_from_snapshot(&self) {
-        if self.accounts_update_notifier.is_none() {
-            return;
-        }
-
-        let mut slots = self.storage.all_slots();
-        let mut notified_accounts: HashSet<Pubkey> = HashSet::default();
-        let mut notify_stats = GeyserPluginNotifyAtSnapshotRestoreStats::default();
+        if let Some(accounts_update_notifier) = &self.accounts_update_notifier {
```
Comment (KirillLykov) [inline accounts-db/src/accounts_db/geyser_plugin_utils.rs]: For the sake of early exit, please use let-else:
```rust
let Some(accounts_update_notifier) = self.accounts_update_notifier.as_ref() else {
    return;
}
```

File: accounts-db/src/accounts_db/geyser_plugin_utils.rs [hunk 2]
```diff
@@ -39,22 +39,25 @@ impl AccountsDb {
...
+        if accounts_update_notifier.notify_snapshot_disabled() {
+            accounts_update_notifier.notify_end_of_restore_from_snapshot();
+        } else {
+            let mut slots = self.storage.all_slots();
+            let mut notified_accounts: HashSet<Pubkey> = HashSet::default();
+            let mut notify_stats = GeyserPluginNotifyAtSnapshotRestoreStats::default();
+
+            slots.sort_by(|a, b| b.cmp(a));
+            for slot in slots {
+                self.notify_accounts_in_slot(slot, &mut notified_accounts, &mut notify_stats);
```
Comment (KirillLykov) [inline accounts-db/src/accounts_db/geyser_plugin_utils.rs:53]: Do you know why this code take minutes?

File: accounts-db/src/accounts_db/geyser_plugin_utils.rs [hunk 3]
```diff
@@ -39,22 +39,25 @@ impl AccountsDb {
...
             return;
-        }
-
-        let mut slots = self.storage.all_slots();
-        let mut notified_accounts: HashSet<Pubkey> = HashSet::default();
-        let mut notify_stats = GeyserPluginNotifyAtSnapshotRestoreStats::default();
+        };
+
+        if accounts_update_notifier.notify_snapshot_disabled() {
+            accounts_update_notifier.notify_end_of_restore_from_snapshot();
```
Comment (KirillLykov) [inline accounts-db/src/accounts_db/geyser_plugin_utils.rs]: What if we `notify_stats.report();` even in the case when disabled? It should report 0s I guess but it is fine. My assumption is that this report is almost zero cost. 
If so, we could reogranize the code like the following:
```rust
if !accounts_update_notifier.notify_snapshot_disabled() {
         let mut slots = self.storage.all_slots();
            let mut notified_accounts: HashSet<Pubkey> = HashSet::default();
            let mut notify_stats = GeyserPluginNotifyAtSnapshotRestoreStats::default();

            slots.sort_by(|a, b| b.cmp(a));
            for slot in slots {
                self.notify_accounts_in_slot(slot, &mut notified_accounts, &mut notify_stats);
            }
}
accounts_update_notifier.notify_end_of_restore_from_snapshot();
notify_stats.report();
```

**Comment by steviez** [comment]
Granted nobody ever reviewed this PR, but this is looking to accomplish the same thing that https://github.com/anza-xyz/agave/pull/2022 does right ?

File: validator/src/cli.rs [hunk 4]
```diff
@@ -1201,6 +1201,13 @@ pub fn app<'a>(version: &'a str, default_args: &'a DefaultArgs) -> App<'a, 'a> {
                 .takes_value(false)
                 .help("Еnable Geyser interface even if no Geyser configs are specified."),
         )
+        .arg(
+            Arg::with_name("geyser_plugin_snapshot_disabled")
```
Comment (lijunwangs) [inline validator/src/cli.rs]: I think better name is geyser_plugin_disable_snapshot_nottification

File: accounts-db/src/accounts_db/geyser_plugin_utils.rs (hunk 2 above)
Comment (lijunwangs) [reply accounts-db/src/accounts_db/geyser_plugin_utils.rs:53]: because of amount of data -- could be 100 millions of records

**Comment by fanatid** [comment]
@steviez oh wow, I remember that PR and my PR doing the same just in different way! Actually, I think that #2022 way to define snapshot requirement through an additional method looks much better than the flag.

@steviez @lijunwangs what do you think, if:
- we remove CLI flag `--geyser-plugin-snapshot-disabled` from this PR
- add function `account_data_snapshot_notifications_enabled` (should we add default value? `true` for backward compatibility?)
- if any plugin return `true` from new function then we stream snapshot data, otherwise skip (as with flag in this PR)

**Comment by fanatid** [comment]
Updated from CLI argument to Geyser interface function

**Comment by lijunwangs** [comment]
> @steviez oh wow, I remember that PR and my PR doing the same just in different way! Actually, I think that #2022 way to define snapshot requirement through an additional method looks much better than the flag.
> 
> @steviez @lijunwangs what do you think, if:
> 
> * we remove CLI flag `--geyser-plugin-snapshot-disabled` from this PR
> * add function `account_data_snapshot_notifications_enabled` (should we add default value? `true` for backward compatibility?)
> * if any plugin return `true` from new function then we stream snapshot data, otherwise skip (as with flag in this PR)

That sounds okay -- so we allow the plugin to opt in.

File: geyser-plugin-manager/src/geyser_plugin_service.rs [hunk 5]
```diff
@@ -90,6 +90,8 @@ impl GeyserPluginService {
 
         let account_data_notifications_enabled =
             plugin_manager.account_data_notifications_enabled() || geyser_plugin_always_enabled;
+        let account_data_snapshot_notifications_enabled =
+            plugin_manager.account_data_notifications_enabled();
```
Comment (lijunwangs) [inline geyser-plugin-manager/src/geyser_plugin_service.rs]:
```suggestion
            plugin_manager.account_data_snapshot_notifications_enabled();
```

Comment (fanatid) [reply geyser-plugin-manager/src/geyser_plugin_service.rs]: Thank you!

**Review by lijunwangs — APPROVED** [review]

**Comment by fanatid** [comment]
@lijunwangs thank you! is it possible to create a backport to v2.1?

**Comment by lijunwangs** [comment]
> @lijunwangs thank you! is it possible to create a backport to v2.1?

We normally narrow the back ports to really critical issues such as security or serious performance problems. Not new features or improvements.

**Comment by paultoback** [comment]
How do I use this? I built from the specific commit but --geyser-plugin-snapshot-disabled or --skip-notify-account-restore-from-snapshot doesn't still seem to exist @fanatid @lijunwangs  - please help, super appreciated!

**Comment by steviez** [comment]
> How do I use this? I built from the specific commit but --geyser-plugin-snapshot-disabled or --skip-notify-account-restore-from-snapshot doesn't still seem to exist

@paultoback - The PR was updated to avoid adding flags; see https://github.com/anza-xyz/agave/pull/4103#issuecomment-2546539134

**Pull request merged**